### ISSUE TITLE
For targets without `idType`, the winner emission is not deterministic

### ISSUE THREAD

**Issue by dianabarsan** [medic/cht-core#6700]
**Describe the bug**
https://docs.communityhealthtoolkit.org/apps/reference/targets/#targetsjs

Targets will, by default, count a single emission per contact, even if they "apply to" reports. The only way to get a target to count multiple reports per contact is to set its `idType` to `report`. This is not the scope of this issue, though. 
When there are multiple emissions for one contact for one target (multiple reports that match the `appliesIf` and `passesIf` for a single contact), the winning emission (whose reported_date will determine which reporting interval the "contact" gets counted towards the target goal) is not deterministic. 

**To Reproduce**
1. Create a simple target that counts reports, this is an example, but you can use any form you desire:
```
{
  id: 'target_forma',
  translation_key: 'target',
  subtitle_translation_key: 'target',
  type: 'count',
  icon: 'no-icon',
  goal: 16,
  appliesTo: 'reports',
  appliesToType: ['forma'],
  appliesIf: function () {
    return true;
  },
  date: 'reported'
}
```
2. Create a bunch of reports that will match this target, all reports about the same patient. 
It's helpful to generate these in a CSV and upload them with medic-conf because you're going to need to change their reported_date and _id fields. 

Create reports in different reporting periods (different months) and make sure their _id fields are out of sort. When I tested, this was my report list:

| _id | form | reported_date | 
| --- | ---- | ------------- |
| a_report | forma | date in February | 
| b_report | forma | date in January | 
| c_report | forma | date in June | 
| d_report | forma | date in April | 
| e_report | forma | date in May | 

So I'm making a very clear alphabetically sorted list of _id with a completely scrambled list of reported_dates. 

3. Login as an offline user that would sync all these reports. Go to the targets page. 
Check if you get a count on your target.
4. Change your OS clock to be in June (`c_report` is the "latest" of its type, so theoretically it should be the one to generate the winning emission). 
Check your targets page and see that you don't have a target count. 
5. Change your OS clock to be in January (`a_report` is the "first" of it's type, so if the first chronological report is selected, this should be the one to generate the winning emission). 
Check your targets page and see that you don't have a target count. 
6. Change your OS clock to Feb / April and see that you don't have a target count. 
7. Change your OS clock to May and see that you DO have a target count. 

**Expected behavior**
I believe, pre 3.8 the last chronological report was generating the winning emission. 
Pre 3.8, we were using the Search service to get a list of reports:
https://github.com/medic/cht-core/blob/3.7.x/webapp/src/js/services/rules-engine.js#L224
Relevant part of the search service is here: https://github.com/medic/cht-core/blob/3.7.x/shared-libs/search/src/search.js#L33 where we sort the results by view value, which is the reported_date: https://github.com/medic/cht-core/blob/3.7.x/ddocs/medic-client/views/reports_by_validity/map.js#L3

Post 3.8 we just query the `reports_by_subject` here https://github.com/medic/cht-core/blob/master/shared-libs/rules-engine/src/pouchdb-provider.js#L34 and never sort the results. 
CouchDB/PouchDB view queries will return rows sorted by the emitted key, and if the key is the same, sorted by _id. The emitted value - which is the reported_date in this case - is not taken into account. 
https://docs.couchdb.org/en/stable/api/ddoc/views.html?highlight=view#api-ddoc-view-sorting

**Environment**
- Instance: any 3.8+
- App: webapp
- Version: 3.8+

**Additional context**
This was reported as a regression by a partner after an upgrade to 3.9.1. 
Not sure about labeling and whether we should consider this production-urgent or not.

**Comment by dianabarsan** [comment]
Ready for AT on `6700-targets-for-reports`

**Comment by newtewt** [comment]
The fix is working to me @dianabarsan.  Spent time creating some data as described above.   Reports with `patient_id` associated to the same person. `_id` was alphabetical. Setting the time back to different months the value is calculated at 0. While setting it to current date the count was 1.  Also confirmed that percent based targets are working, targets for contacts are working. 

If we are back porting this just give me the details and I'd be glad to jump on it. Attached the sample csvs I used for testing down the road. 

 ``` csv
reference_id:excluded,parent:place WHERE reference_id=COL_VAL,name,phone,sex,role,reported_date,patient_id
p_hc1,health_center_1,Bob Johnson 1,16145558771,male,manager,1552494835669,60951

```

``` csv
reference_id:excluded,parent,is_name_generated,name,reported_date:timestamp
district_1,,false,Pregnancies,1544031155715
```

``` csv 
reference_id:excluded,parent:place WHERE reference_id=COL_VAL,is_name_generated,name,reported_date:timestamp
health_center_1,district_1,FALSE,HC1,1544031155715
```

``` csv
"content_type",reported_date:timestamp,"fields.inputs.source","fields.inputs.contact._id","patient_id:GET patient_id OF person WHERE reference_id=COL_VAL"
"xml",1604331091848,"contact","6b77d196-0ac8-4014-a2f5-8edf0a074c50","p_hc1"
"xml",1604331091848,"contact","6b77d196-0ac8-4014-a2f5-8edf0a074c50","p_hc1"
"xml",1580658669000,"contact","6b77d196-0ac8-4014-a2f5-8edf0a074c50","p_hc1"
"xml",1583164269000,"contact","6b77d196-0ac8-4014-a2f5-8edf0a074c50","p_hc1"
"xml",1585842669000,"contact","6b77d196-0ac8-4014-a2f5-8edf0a074c50","p_hc1"
"xml",1588434669000,"contact","6b77d196-0ac8-4014-a2f5-8edf0a074c50","p_hc1"
"xml",1593705069000,"contact","6b77d196-0ac8-4014-a2f5-8edf0a074c50","p_hc1"
"xml",1596383469000,"contact","6b77d196-0ac8-4014-a2f5-8edf0a074c50","p_hc1"
"xml",1599061869000,"contact","6b77d196-0ac8-4014-a2f5-8edf0a074c50","p_hc1"
"xml",1601653869000,"contact","6b77d196-0ac8-4014-a2f5-8edf0a074c50","p_hc1"
"xml",1604331091848,"contact","6b77d196-0ac8-4014-a2f5-8edf0a074c50","p_hc1"
```

```csv
username,password,roles,phone,contact:person WHERE reference_id=COL_VAL,place:GET _id OF place WHERE reference_id=COL_VAL
ac1,Secret_1,district_admin:red1,16145558779,p_hc1,health_center_1

```

**Comment by dianabarsan** [comment]
Thanks @newtewt . 
@garethbowen , please advise in regards to backporting. Thanks!

**Comment by garethbowen** [comment]
@dianabarsan I think we should backport to 3.10.x and 3.9.x. Could you please do the backport to those branches? I can look after the releases from there.

**Comment by dianabarsan** [comment]
Merged into `master` and back-ported to `3.10.x` and `3.9.x`.
